### ISSUE TITLE
EfficientNet for Face Recognition

### ISSUE THREAD

**Issue by radu-diaconescu13** [tensorflow/tpu#464]
Hello,

First of all thank you for your hard work on this repo.
I am trying to use Efficient Net as an embedding extractor for face recognition.
I am using arc face loss to train but I get very poor face verification performance compared to MobileNet v2 on the same training set for example. 
Do you have any suggestion for training?
Thank you

**Comment by CretuCalin** [comment]
I'm also interested in finding out.

**Comment by saberkun** [comment]
Do you use the same training scheme as we post like applying moving average?

**Comment by nmber5** [comment]
> Hello,
> 
> First of all thank you for your hard work on this repo.
> I am trying to use Efficient Net as an embedding extractor for face recognition.
> I am using arc face loss to train but I get very poor face verification performance compared to MobileNet v2 on the same training set for example.
> Do you have any suggestion for training?
> Thank you

have you get a better results, i will try the same thing as you?

**Comment by radu-diaconescu13** [comment]
Hello. No, I have not. I tried taking a smaller part of B-0 network, but still got very poor results.

**Comment by nmber5** [comment]
> Hello @nmber5 . No, I have not. I tried taking a smaller part of B-0 network, but still got very poor results.
> I got 25TPR versus 85TPR with a mobilenet V2 encoder

it's really poor

**Comment by mingxingtan** [comment]
Hi @radu-diaconescu13 and @nmber5,

I am not familiar to face recognition models, but here are some possible aspects you may want to take a look:

(1) the default settings are for training ImageNet, which may not be good for face recognition models, you may need to use the same training settings as mobilenetv2. In particular, you might want to set [drop_connect_rate=0 ](https://github.com/tensorflow/tpu/blob/5a5d62339b08eb4366515c05e405104ab79297a4/models/official/efficientnet/efficientnet_builder.py#L176) and [dropout_rate=0](https://github.com/tensorflow/tpu/blob/5a5d62339b08eb4366515c05e405104ab79297a4/models/official/efficientnet/efficientnet_builder.py#L38). You may also want to adjust weight_decay, learning_rate, and so on.

(2) If you still cannot get good results and cannot find out the issue, you can do a sanity check by building mobilenetv2 using this efficientnet codebase. Basically, you just need to have another [block_args](https://github.com/tensorflow/tpu/blob/5a5d62339b08eb4366515c05e405104ab79297a4/models/official/efficientnet/efficientnet_builder.py#L166) for mobilenetv2.

(3) Lastly, I want to point out that EfficientNet-B0 is still slightly bigger than mobilenetv2. Although bigger models usually lead to better performance, it may not be true for your case. You may consider train your model with mobienetv2(depth_multiplier=1.4x) to see whether its performance is better than the basic mobilenetv2.

Hope it helps.

**Comment by radu-diaconescu13** [comment]
Thank you @mingxingtan  for you answer.

I tried already (1) but with no luck.  (2) and (3) are very good ideas.
I will try them in the future as at the moment i am not working on face recognition anymore.

Thank you again

**Comment by nmber5** [comment]
> Hello,
> 
> First of all thank you for your hard work on this repo.
> I am trying to use Efficient Net as an embedding extractor for face recognition.
> I am using arc face loss to train but I get very poor face verification performance compared to MobileNet v2 on the same training set for example.
> Do you have any suggestion for training?
> Thank you

@radu-diaconescu13  Do you train on multi GPUs or TPU？When i train on gpus, the train code cna find more then one visible gpus, but only use the first one(device:0). I tried to set session options but did not work for me.
 @radu-diaconescu13 and @mingxingtan how can i training with mutli gpus? Can you guys give me some help?
thanks a lot

**Comment by zeakey** [comment]
> > Hello,
> > First of all thank you for your hard work on this repo.
> > I am trying to use Efficient Net as an embedding extractor for face recognition.
> > I am using arc face loss to train but I get very poor face verification performance compared to MobileNet v2 on the same training set for example.
> > Do you have any suggestion for training?
> > Thank you
> 
> have you get a better results, i will try the same thing as you?

I encountered the same problem recently. I use the pytorch implementation of EfficientNet from the [timm](https://github.com/rwightman/pytorch-image-models/tree/master/timm) package.

**Comment by zeakey** [comment]
@mingxingtan Could you please provide the [blocks_args](https://github.com/tensorflow/tpu/blob/5a5d62339b08eb4366515c05e405104ab79297a4/models/official/efficientnet/efficientnet_builder.py#L166) for MobileNetV2, or is there any docs for the formatting of block_args?
